In [1]:

#Importar numpy y matplotlib.pyplot
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import scipy.optimize as opt

# Proyecto Inversión de Portafolio en la Bolsa 

## Objetivo: Maximizar el retorno de inversión al final del periodo de 6 años
### Realizar de manera gráfica el comportamiento de los instrumentos de inversión en cuanto a su tipo, su periodo de vencimiento y su disponibilidad
### Obtener la cantidad de dinero específica que se debe colocar en cada instrumento de inversión durante el periodo de los 6 años. 


## Problema: 


Una casa de bolsa ha reunido un capital de **600 millones de pesos para administrarlo por seis años**, por lo que está interesada en encontrar el mejor portafolio de inversiones que le permita maximizar el rendimiento de la inversión en ese periodo. Para este fin, se estudiaron algunos instrumentos de inversión a plazo fijo, obteniéndose datos del rendimiento ofrecido al final del vencimiento. Los periodos de vencimiento y la disponibilidad del instrumento para invertir en él. Al analizar los instrumentos de inversión se vio la necesidad de tener una alternativa más que comprendiera el dinero que intencionalmente no se invierta en ese año esperando una mejor opción de inversión en un periodo futuro. Obviamente el rendimiento de esta alternativa es nulo, no tiene vencimiento y está disponible en forma inmediata, es decir cada año. Con los datos recabados se construyó la siguiente tabla: 

Instrumento de Inversion|Rendimiento al Final del Periodo (%)|Periodo de Vencimiento (años)|Disponibilidad del Instrumento 
:----|----|----|----
Tipo 1|28|3|Cada Año
Tipo 2|16|2|Cada Año
Tipo 3|50|3|Principios de año 2
Tipo 4|40|2|Cada año después del año 3
Tipo 5|45|4|Principios del año 1
Tipo 6|0|1|Cada año inmediato

La casa de bolsa quiere desarrollar un modelo que le permita determinar el portafolio financiero que maximice el rendimiento del paquete completo al final del año 6.

### Declaración de variables 
- $T_{11}$=Intrumento tipo 1 disponibilididad año 0 vencimineto año 3
- $T_{12}$=Intrumento tipo 1 disponibilididad año 1 vencimineto año 4
- $T_{13}$=Intrumento tipo 1 disponibilididad año 2 vencimineto año 5
- $T_{14}$=Intrumento tipo 1 disponibilididad año 3 vencimineto año 6
- $T_{21}$=Intrumento tipo 2 disponibilididad año 0 vencimineto año 2
- $T_{22}$=Intrumento tipo 2 disponibilididad año 1 vencimineto año 3
- $T_{23}$=Intrumento tipo 2 disponibilididad año 2 vencimineto año 4
- $T_{24}$=Intrumento tipo 2 disponibilididad año 3 vencimineto año 5
- $T_{25}$=Intrumento tipo 2 disponibilididad año 4 vencimineto año 6
- $T_{31}$=Intrumento tipo 3 disponibilididad año 2 vencimineto año 5
- $T_{41}$=Intrumento tipo 4 disponibilididad año 3 vencimineto año 5
- $T_{42}$=Intrumento tipo 4 disponibilididad año 4 vencimineto año 6
- $T_{51}$=Intrumento tipo 5 disponibilididad año 1 vencimineto año 5
- $T_{61}$=Intrumento tipo 6 disponibilididad año 0 vencimineto año 1
- $T_{62}$=Intrumento tipo 6 disponibilididad año 1 vencimineto año 2
- $T_{63}$=Intrumento tipo 6 disponibilididad año 2 vencimineto año 3
- $T_{64}$=Intrumento tipo 6 disponibilididad año 3 vencimineto año 4
- $T_{65}$=Intrumento tipo 6 disponibilididad año 4 vencimineto año 5
- $T_{66}$=Intrumento tipo 6 disponibilididad año 5 vencimineto año 6

#### *Función a Max* 

$ \max_{T_{11}...T_{66}}F.O.=1.28T_{14}+1.16T_{25}+1.4T_{42}+T_{66}$

*S.A* 
+ Año1: $T_{11}+T_{21}+T_{51}+T_{61} \leq 600,000,000$
+ Año2: $T_{12}+T_{22}+T_{31}+T_{62}=T_{61}$
+ Año3: $T_{13}+T_{23}+T_{63}=1.16T_{21}+T_{62}$
+ Año4: $T_{14}+T_{24}+T_{41}+T_{64}=1.16T_{22}+T_{63}+1.28T_{11}$
+ Año5: $T_{25}+T_{42}+T_{65}=1.28T_{12}+1.16T_{23}+1.5T_{31}+1.45T_{51}+T_{64}$
+ Año6: $T_{66}=1.28T_{13}+1.16T_{24}+1.4T_{41}+T_{65}$


* La función objetivo representa la manera en la que se debe administrar el dinero en los diferentes instrumentos durante los 6 años del periodo. Es decir, la función objetivo es maximizar el retorno de los bonos con fin en el periodo del año 6
* Las restricciones están dadas por la disponibilidad de cada bono dependiendo del tiempo, y los bonos que vencen en el mismo periodo, por ejemplo, la restricción de inicios del "Año 4", está limitado por la emisión de los bonos $T_{14}$, $T_{24}$, $T_{41}$ y $T_{61}$, y el vencimiento de los bonos con retorno de $1.16T_{22}$, $T_{63}$, $1.28T_{11}$, igualamos por que lo que entra es igual a lo que sale multiplicado siempre por el retorno. Se puede observar de mejor manera en la siguiente tabla:

### Explicación gráfica de las restricciones, hay que tener en cuenta el rendimiento de cada instrumento al finalizar su periodo de vencimiento.

![.](GRAF.PNG)

In [2]:
c = -np.array([0,0,0,1.28, 0,0,0,0,1.16, 0,0,1.4, 0,0,0,0,0,0,1])

A = ([[1,0,0,0,1,0,0,0,0,0,0,0,1,1,0,0,0,0,0]])

b = np.array([600])

A_eq = np.array([[0,1,0,0,0,1,0,0,0,1,0,0,0,-1,1,0,0,0,0],
                [0,0,1,0,-1.16,0,1,0,0,0,0,0,0,0,-1,1,0,0,0],
                [-1.28,0,0,1,0,-1.16,0,1,0,0,1,0,0,0,0,-1,1,0,0],
                [0,-1.28,0,0,0,0,-1.16,0,1,-1.5,0,1,-1.45,0,0,0,-1,1,0],
                [0, 0,-1.28,0,0,0,0,-1.16,0,0,-1.4,0,0,0,0,0,0,-1,1]])

b_eq = np.array([0,0,0,0,0])


In [3]:
sol2 = opt.linprog(c, A, b, A_eq, b_eq, method="simplex")
sol2

     con: array([0., 0., 0., 0., 0.])
     fun: -1260.0
 message: 'Optimization terminated successfully.'
     nit: 10
   slack: array([0.])
  status: 0
 success: True
       x: array([  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0., 600.,   0.,
       900.,   0., 600.,   0.,   0.,   0.,   0.,   0.])

![.](RESP.PNG)

### *Conclusiones*:
Los resultados obtenidos nos muestran cómo es que se debe de distribuir el capital a lo largo de los 6 aós del periodo de inversión, invirtiendo primero los 600 millones en el primer bono del **Tipo 6 (T61)**, finalizando el periodo de un año invertir los mismos 600 millones (ya que el instrumento de Tipo 6 no tiene retorno) en el instrumento de **Tipo 3 (T31)**, al finalizar el periodo a inicios de año 5 mover los 900 millones (600 millones más el retorno de inversión del instrumento de Tipo 3 del 50%) al último instrumento disponible del **Tipo 4 (T42)**, dejándonos una suma final del periodo de inversión de un total de **1260 millones de pesos** al finalizar el periodo de 6 años.

### Referencias:
A. Vega (2019): Investigación de operaciones. ITESO